### PR TITLE
#1304: Replace LangChain Slack helpers with slack-sdk

### LINKED ISSUE #1304: Remove langchain-community from Slack tool
Part of #1242.

`coded_tools/tools/agentic_rag/slack.py` uses `langchain_community.tools.slack.get_channel` / `get_message`. A replacement should use `slack_sdk` directly.

Requirement: keep the fallback-to-mock-data behavior scoped to the "slack-sdk not installed / not configured" case — don't catch broad `Exception`, which would mask real API errors as demo data.

### PR DESCRIPTION
**Pull request by arnabnandy7**
## Description

Replaces the LangChain Community Slack helpers used by the agentic RAG Slack tool with the asynchronous `slack-sdk` client.

Mock data remains available only when `slack-sdk` is unavailable or `SLACK_BOT_TOKEN` is not configured. Errors returned by the Slack API are allowed to propagate instead of being masked by demo data.

Fixes #1304

## Impact

- Removes the Slack tool’s dependency on `langchain_community.tools.slack`.
- Uses `AsyncWebClient` to list channels and retrieve message history.
- Preserves existing responses for missing channels and demo environments.
- Adds `slack-sdk` as a direct dependency.
- Adds unit coverage for configured Slack access, fallback behavior, missing channels, input validation, and API error propagation.

## Screenshots / Logs / Examples (optional)

Not applicable.

## Type of Change

- [x] Bug fix
- [ ] New feature
- [ ] Dependency upgrade
- [x] Refactoring / Improvement
- [ ] Documentation
- [ ] New agent / tool

## Checklist

- [ ] Tested locally
- [x] Added/updated tests
- [ ] Updated relevant documentation
- [x] Added dependencies to appropriate `requirements.txt` (tool-specific preferred; main only for core functionality)

---

**By submitting this pull request, I confirm that my contribution is made under the terms of the project's [Apache 2.0 License](../LICENSE.txt).**

### CODE REVIEW

**Comment by arnabnandy7** [comment]
@Noravee for your review.

File: requirements.txt [hunk 1]
```diff
@@ -31,6 +31,9 @@ langchain-google-genai>=4.2.2,<5.0
 
 langchain-community>=0.4.2
 
+# For reading Slack channels in the agentic RAG demo
+slack-sdk>=3.39.0,<4.0
```
Comment (Noravee) [inline requirements.txt]: `slack-sdk` shouldn't be a hard dependency. This tool backs the `agentic_rag` demo network, which ships disabled by default in `registries/tools/manifest.hocon`, and the `try/except ImportError` fallback you kept exists precisely so the package can be absent.

Please remove it from `requirements.txt` and document the requirement where users opt in instead:

- a comment on the `slack_tool` entry in `registries/tools/agentic_rag.hocon` (install `slack-sdk` and set `SLACK_BOT_TOKEN` to use this tool)
- a comment on the `"tools/agentic_rag.hocon"` line in `registries/tools/manifest.hocon`
- a short setup note in `docs/examples/tools/agentic_rag.md`

(This is also what the PR checklist means by "tool-specific preferred; main only for core functionality".)

File: coded_tools/tools/agentic_rag/slack.py [hunk 2]
```diff
@@ -83,34 +82,26 @@ async def async_invoke(self, args: Dict[str, Any], sly_data: Dict[str, Any]) ->
...
-            # Get channel id and return the messages if possible.
-            channel_id: str = channel_id_name_dict.get(channel_name)
-            if channel_id:
-                return await SlackGetMessage().ainvoke(channel_id)
+        slack_bot_token = os.getenv("SLACK_BOT_TOKEN")
+        if AsyncWebClient is None or not slack_bot_token:
+            return self._mock_response(channel_name)
+
+        client = AsyncWebClient(token=slack_bot_token)
+        channel_response = await client.conversations_list()
```
Comment (Noravee) [inline coded_tools/tools/agentic_rag/slack.py]: Two regressions here vs. the langchain `SlackGetChannel` this replaces:

1. **Private channels are no longer found.** The old tool fetched `public_channel` plus `private_channel,mpim,im`; `conversations_list()` with no args returns public channels only, so a private channel name now comes back as "not found".
2. **No pagination.** The old tool looped over `response_metadata.next_cursor` with `limit=200` until exhausted; a single call returns at most one page (default 100), so workspaces with many channels can miss the channel entirely.

Please pass `types="public_channel,private_channel"` and paginate with the cursor.

File: coded_tools/tools/agentic_rag/slack.py [hunk 3]
```diff
@@ -83,34 +82,26 @@ async def async_invoke(self, args: Dict[str, Any], sly_data: Dict[str, Any]) ->
...
+        channel_id_name_dict = {channel["name"]: channel["id"] for channel in channel_response["channels"]}
+
+        channel_id = channel_id_name_dict.get(channel_name)
+        if not channel_id:
             return f"The {channel_name} channel not found."
 
-        except json.JSONDecodeError as err:
-            return f"Error: Could not parse slack channel list: {err}"
+        message_response = await client.conversations_history(channel=channel_id)
+        return json.dumps(message_response["messages"])
```
Comment (Noravee) [inline coded_tools/tools/agentic_rag/slack.py]: The old `SlackGetMessage` filtered each message down to `{user, text, ts}`. Returning the raw `messages` payload hands the LLM full Slack message objects (blocks, reactions, team/client ids, …), which bloats the model's context and includes metadata the agent doesn't need. Please restore the field filtering.

File: coded_tools/tools/agentic_rag/slack.py [hunk 4]
```diff
@@ -83,34 +82,26 @@ async def async_invoke(self, args: Dict[str, Any], sly_data: Dict[str, Any]) ->
...
-            # Convert the JSON str to a list
-            channel_id_name_list: list = json.loads(channel_id_name_str)
-            # Make a lookup table with channel names as keys and ids as values
-            channel_id_name_dict: dict = {channel["name"]: channel["id"] for channel in channel_id_name_list}
-
-            # Get channel id and return the messages if possible.
-            channel_id: str = channel_id_name_dict.get(channel_name)
-            if channel_id:
-                return await SlackGetMessage().ainvoke(channel_id)
+        slack_bot_token = os.getenv("SLACK_BOT_TOKEN")
```
Comment (Noravee) [inline coded_tools/tools/agentic_rag/slack.py]: The langchain `login()` accepted `SLACK_BOT_TOKEN` *or* `SLACK_USER_TOKEN`. Only checking `SLACK_BOT_TOKEN` means anyone currently configured with a user token silently gets mock demo data back — which looks like success. Please either check both env vars or call the change out explicitly in the docs.

**Review by Noravee — CHANGES_REQUESTED** [review]
Thanks — this is scoped exactly as asked (one tool, its own PR against #1304), and I'm glad to see API errors now propagate with a test proving it.

A few changes are needed before merge though. The main one is dependency placement, plus a few behavior regressions vs. the langchain tool this replaces (I compared against the `langchain-community` 0.4.2 source). See the inline comments.

Also, the branch is behind `main` now that #1247 merged — please rebase/update when you push the fixes.

**Comment by arnabnandy7** [comment]
@Noravee all taken care off. Please proceed with review once the CI passes.

File: coded_tools/tools/agentic_rag/slack.py [hunk 5]
```diff
@@ -83,34 +82,35 @@ async def async_invoke(self, args: Dict[str, Any], sly_data: Dict[str, Any]) ->
...
+            if not cursor:
+                break
+
+        if not channel_id:
             return f"The {channel_name} channel not found."
 
-        except json.JSONDecodeError as err:
-            return f"Error: Could not parse slack channel list: {err}"
+        message_response = await client.conversations_history(channel=channel_id)
+        messages = [{key: message[key] for key in ("user", "text", "ts")} for message in message_response["messages"]]
```
Comment (Noravee) [inline coded_tools/tools/agentic_rag/slack.py]: Thanks for the quick turnaround — everything from the review is addressed, and the added pagination and user-token tests look great. One last edge case before I approve:

`message[key]` raises `KeyError` for messages that don't carry all three fields — messages posted by incoming webhooks and classic bots have `bot_id` instead of `user`, so a single app-posted message in the channel makes the whole tool call fail. The old langchain tool guarded against exactly this by skipping such messages:

```suggestion
        messages = [
            {key: message[key] for key in ("user", "text", "ts")}
            for message in message_response["messages"]
            if all(key in message for key in ("user", "text", "ts"))
        ]
```

Comment (arnabnandy7) [reply coded_tools/tools/agentic_rag/slack.py]: Done

**Review by Noravee — APPROVED** [review]
LGTM.
